src/protocol/ipa/mod.rs

```diff
@@ -0,0 +1,200 @@
+            Self::ApplySortPermutation => "apply_sort_permutation",
+            Self::ComputeHelperBits => "compute_helper_bits",
+            Self::AccumulateCredit => "accumulate_credit",
+            Self::PerformUserCapping => "user_capping",
+        }
+    }
+}
+
+/// # Errors
+/// Propagates errors from multiplications
+#[allow(dead_code)]
+pub async fn ipa<F>(
+    ctx: SemiHonestContext<'_, F>,
+    mk_shares: &[XorReplicated],
+    num_bits: u32,
+    other_inputs: Vec<Vec<Replicated<F>>>,
+    per_user_credit_cap: u32,
+) -> Result<Vec<Replicated<F>>, Error>
+where
+    F: Field,
+{
+    debug_assert_eq!(mk_shares.len(), other_inputs.len());
+
+    let local_lists = convert_all_bits_local(ctx.role(), mk_shares, num_bits);
+    let converted_shares = convert_all_bits(
+        &ctx.narrow(&Step::ModulusConversionForMatchKeys),
+        &local_lists,
+    )
+    .await
+    .unwrap();
+    let sort_permutation = generate_permutation(
+        ctx.narrow(&Step::GenSortPermutationFromMatchKeys),
+        &converted_shares,
+        num_bits,
+    )
+    .await
+    .unwrap();
+
+    let num_bits: usize = usize::try_from(num_bits).unwrap();
+
+    let combined_match_keys_and_sidecar_data = other_inputs
+        .into_iter()
+        .enumerate()
+        .map(|(row_index, mut sidecar_data)| {
+            for i in 0..num_bits {
+                sidecar_data.push(converted_shares[i][row_index].clone());
+            }
+            sidecar_data
+        })
+        .collect::<Vec<_>>();
+
+    let sorted_rows = sort_permutation
+        .apply(
+            ctx.narrow(&Step::ApplySortPermutation),
+            combined_match_keys_and_sidecar_data,
+        )
+        .await
+        .unwrap();
+
+    let futures = zip(
+        repeat(ctx.narrow(&Step::ComputeHelperBits)),
+        sorted_rows.iter(),
+    )
+    .zip(sorted_rows.iter().skip(1))
+    .enumerate()
+    .map(|(i, ((ctx, row), next_row))| {
+        let record_id = RecordId::from(i);
+        async move { bitwise_equal(ctx, record_id, &row[3..], &next_row[3..]).await }
+    });
+    let helper_bits = try_join_all(futures).await?;
+
+    let attribution_input_rows = sorted_rows
+        .iter()
+        .enumerate()
+        .map(|(i, row)| {
+            let hb = if i == 0 {
+                Replicated::ZERO
+            } else {
+                helper_bits[i - 1].clone()
+            };
+            AttributionInputRow {
+                is_trigger_bit: row[0].clone(),
+                helper_bit: hb,
+                breakdown_key: row[1].clone(),
+                credit: row[2].clone(),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let accumulated_credits =
+        accumulate_credit(ctx.narrow(&Step::AccumulateCredit), &attribution_input_rows).await?;
+
+    let _user_capped_credits = credit_capping(
+        ctx.narrow(&Step::PerformUserCapping),
+        &accumulated_credits,
+        per_user_credit_cap,
+    )
+    .await?;
+
```

Comment:
We still need to call the next protocols:
1. Aggregate down to sums per breakdown key
2. Add random noise for DP